src/main.ts

```diff
@@ -20,48 +17,74 @@ export default class FormatNotionPlugin extends Plugin {
 			],
 		});
 
-		// Editor mode (right click on text)
 		this.registerEvent(this.app.workspace.on('editor-menu',
-			(menu: Menu, _: Editor, view: MarkdownView) => {
-				const url = this.getCleanedUrl();
-				if (url) {
-					menu.addItem((item) => {
+			(menu: Menu, editor: Editor, view: MarkdownView) => {
+				const urlInfo = this.getYoutubeUrl(editor);
+				if (urlInfo) {
+					menu.addItem((item) => { 
 						item.setTitle("Url to Preview/Iframe")
 							.setIcon("create-new")
 							.onClick((_) => {
-								this.urlToIframe(url);
+								this.urlToIframe(urlInfo, editor);
 							});
 					});
 				}
 			}));
-
 	}
 
-	async urlToIframe(inputUrl?: string): Promise<void> {
-		const activeLeaf: any = this.app.workspace.activeLeaf;
-		const editor = activeLeaf.view.sourceMode.cmEditor;
+	async urlToIframe(inputUrlInfo?: {url: string, start: EditorPosition, end: EditorPosition}, inputEditor?: Editor): Promise<void> {
+		const activeLeaf = this.app.workspace.activeLeaf;
+		const editor = inputEditor || (activeLeaf.view as MarkdownView).editor;
 
-		const url = inputUrl || this.getCleanedUrl()
+		const urlInfo = inputUrlInfo || this.getYoutubeUrl(editor);
 
-		if (url) {
-			const iframeHtml = await getIframe(url)
-			const modal = new ConfigureIframeModal(this.app, iframeHtml, editor)
+		if (urlInfo) {
+			editor.setSelection(urlInfo.start, urlInfo.end);
+			const iframeHtml = await getIframe(urlInfo.url);
+			const modal = new ConfigureIframeModal(this.app, iframeHtml, editor);
+			modal.onClose = () => {
+				if (modal.iframeHtml) {
+					// Ensure we're still selecting the correct range
+					const currentContent = editor.getRange(urlInfo.start, urlInfo.end);
+					if (currentContent === urlInfo.url) {
+						// Replace the URL with the iframe HTML
+						editor.replaceRange(modal.iframeHtml, urlInfo.start, urlInfo.end);
+						new Notice('YouTube URL replaced with iframe.');
+					} else {
+						new Notice('URL content changed, unable to replace.');
+					}
+				}
+			};
 			modal.open();
 		} else {
-			new Notice('Select a URL to convert to an preview/iframe.');
+			new Notice('No valid YouTube URL found at cursor position or in selection.');
 		}
 	}
 
-	private getCleanedUrl(): string {
-		const activeLeaf: any = this.app.workspace.activeLeaf;
-		const editor = activeLeaf.view.sourceMode.cmEditor;
-		const selectedText: string = editor.somethingSelected() ? editor.getSelection() : null;
-		const cleanedText = selectedText?.trim();
+	private getYoutubeUrl(editor: Editor): {url: string, start: EditorPosition, end: EditorPosition} | null {
+		const cursor = editor.getCursor();
+		const line = editor.getLine(cursor.line);
+		
+		const youtubeRegex = /https?:\/\/(?:www\.)?(?:youtube\.com\/watch\?v=|youtu\.be\/)([a-zA-Z0-9_-]+)(?:&\S*)?/;

```

Comment:
Can you make sure that we detect any url? The goal is to support other urls than youtube, for example twitter etc.